client/state/test/data-fetching-persistence.tsx

```diff
@@ -0,0 +1,281 @@
+/**
+ * @jest-environment jsdom
+ */
+
+import '@testing-library/jest-dom/extend-expect';
+import { render, waitFor } from '@testing-library/react';
+import { QueryClient, QueryClientProvider, useQuery, setLogger } from 'react-query';
+import { createWebStoragePersistor } from 'react-query/createWebStoragePersistor-experimental';
+import { persistQueryClient } from 'react-query/persistQueryClient-experimental';
+import { shouldDehydrateQuery } from '../should-dehydrate-query';
+
+const queryClient = new QueryClient();
+
+const PERSISTENCE_KEY = 'REACT_QUERY_OFFLINE_CACHE';
+
+const queryKey = '123';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+const noop = () => {};
+
+// https://github.com/tannerlinsley/react-query/blob/2771a15403cb2e7c70022b850e8c54c6d2b3d8a0/docs/src/pages/guides/testing.md#turn-off-network-error-logging
+setLogger( {
+	error: noop,
+	log: noop,
+	warn: noop,
+} );
+
+class Storage {
+	cache: Map< string, string >;
+
+	constructor() {
+		this.cache = new Map();
+	}
+
+	getItem( key: string ) {
+		return this.cache.get( key );
+	}
+
+	setItem( key: string, value: string ) {
+		this.cache.set( key, value );
+	}
+
+	removeItem( key: string ) {
+		this.cache.delete( key );
+	}
+
+	clear() {
+		this.cache.clear();
+	}
+
+	get length() {
+		return this.cache.size;
+	}
+
+	key() {
+		return 'noop';
+	}
+}
+
+const storage = new Storage();
+
+const getOfflinePersistence = () => JSON.parse( storage.getItem( PERSISTENCE_KEY ) );
+
+const offlinePersistor = createWebStoragePersistor( {
+	storage,
+	key: PERSISTENCE_KEY,
+	throttleTime: 0,
+} );
+
+interface DataFetchingComponentProps< T > {
+	queryFn(): Promise< T >;
+	persistencePredicate?: boolean | ( ( data: T ) => boolean );
+}
+
+const DataFetchingComponent = < T, >( {
+	queryFn,
+	persistencePredicate,
+}: DataFetchingComponentProps< T > ) => {
+	const { status } = useQuery( queryKey, queryFn, {
+		retry: false,
+		meta: {
+			persist: persistencePredicate != null ? persistencePredicate : undefined,
+		},
+	} );
+
+	return <>{ status }</>;
+};
+
+const TestComponent = < T, >( dataFetchingProps: DataFetchingComponentProps< T > ) => {
+	return (
+		<QueryClientProvider client={ queryClient }>
+			<DataFetchingComponent { ...dataFetchingProps } />
+		</QueryClientProvider>
+	);
+};
+
+describe( 'shouldDehydrateQuery', () => {
+	beforeAll( async () => {
+		await persistQueryClient( {
+			queryClient,
+			persistor: offlinePersistor,
+			dehydrateOptions: {
+				shouldDehydrateQuery,
+			},
+		} );
+	} );
+
+	beforeEach( () => {
+		storage.clear();
+		queryClient.clear();
+	} );
+
+	it( 'does not persist failed queries', async () => {
+		const data = 'Hello, World!';
+
+		const queryFn = () => Promise.reject( data );
+
+		const { getByText } = render(
+			<TestComponent queryFn={ queryFn } persistencePredicate={ true } />
```

Comment:
`={ true }` is redundant and can be removed.

```suggestion
			<TestComponent queryFn={ queryFn } persistencePredicate />
```

Reply:
It is, I just wanted to be explicit.

Reply:
I disagree - I prefer using the shorthand because it's shorter, and more straightforward. Ideally, I'd love to see it all throughout the code and enforcing it at some point.

If someone can't read the intention of the shorthand, then they likely won't be able to read the intention of the longer version. 

Anyway, I agree to disagree if you feel strongly about this.

Reply:
No hard feelings :-) Let me change it.